### PR TITLE
Support for advanced generics

### PR DESCRIPTION
**Pull request by alexeev509**
Was added supporting of generic types (declaring and usage):

1. Declaring generic functions and procs.
2. Usage of the generic procs.
3. Declaring generic actions.
4. Usage generic actions.
5. Declaring generic typealias.
6. Creating generic automatons
7. Usage generic pros's of automatons.
8. Creating objects of generic automatons.
9. Usage is/as with generic types.
11. Declaring generic types.
12. Added keywords: in, out, where, ?.

Was added class: class GenericType.
Were added generic test's (see generic.lsl).
Was added field TypeReference in the TypeReference class.

### CODE REVIEW

File: src/test/testdata/expected/lsl/AVLTreeArkTsFunSignatures.lsl [hunk 1]
```diff
@@ -0,0 +1,56 @@
+libsl "1.1.0";
+library std version "11" language "Java" url "https://gitee.com/openharmony/arkcompiler_runtime_core/blob/master/static_core/plugins/ets/stdlib/std/containers/AVLTree.ets";
```
Comment (dog-m) [inline src/test/testdata/expected/lsl/AVLTreeArkTsFunSignatures.lsl]: I strongly suggest removing any mention of third parties, especially trademarks and other IPs (intellectual properties).

```suggestion
library std version "11" language "unknown" url "-";
```

File: src/test/testdata/expected/lsl/generics.lsl [hunk 2]
```diff
@@ -0,0 +1,95 @@
+libsl "1.1.0";
+library std version "11" language "Java" url "-";
```
Comment (dog-m) [inline src/test/testdata/expected/lsl/generics.lsl]:
```suggestion
library std version "11" language "unknown" url "-";
```

File: src/test/testdata/expected/lsl/generics.lsl [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    fun *.genericTypeDefBlockReturnType(x: K): HashMap<Int, Int> {
+        var newHashMap: HashMap<Int, Int> = new A<Int, Int>(state = Initialized);
+        A(newHashMap)._genericProc<Int, Int, Int>();
+        result = newHashMap;
+        action PLAIN_GENERIC_ACTION<Int>(5, 6);
+        if (x is HashMap<Int, Int>) {
+
+        }
+        var obj: HashMap<Int, Int> = x as HashMap<Int, Int>;
+        var newHashMapUnbounded: HashMap<?, ?> = new A<?, ?>(state = Initialized);
```
Comment (dog-m) [inline src/test/testdata/expected/lsl/generics.lsl]: I think, automata constructors should accept only real (i.e. non-wildcard) generic type arguments.

File: src/test/testdata/expected/lsl/generics.lsl [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        A(newHashMapUnbounded)._genericProc<?, ?, ?>();
+        action PLAIN_GENERIC_ACTION<?>(5, 6);
+        if (x is HashMap<?, ?>) {
+
+        }
+        var obj2: HashMap<?, ?> = x as HashMap<?, ?>;
+        var newHashMapBounded: HashMap<in Int, out String> = new A<in Int, out String>(state = Initialized);
```
Comment (dog-m) [inline src/test/testdata/expected/lsl/generics.lsl]: I think, automata constructors should accept only real (i.e. non-wildcard) generic type arguments.

File: src/test/testdata/lsl/AVLTreeArkTsFunSignatures.lsl [hunk 5]
```diff
@@ -0,0 +1,105 @@
+libsl "1.1.0";
+
+library std
+    version "11"
+    language "Java"
+    url "https://gitee.com/openharmony/arkcompiler_runtime_core/blob/master/static_core/plugins/ets/stdlib/std/containers/AVLTree.ets";
```
Comment (dog-m) [inline src/test/testdata/lsl/AVLTreeArkTsFunSignatures.lsl]: I strongly suggest removing any mention of third parties, especially trademarks and other IPs (intellectual properties).
```suggestion
    language "unknown"
    url "-";
```

File: src/test/testdata/lsl/generics.lsl [hunk 6]
```diff
@@ -0,0 +1,118 @@
+///#! pragma: non-synthesizable
+libsl "1.1.0";
+
+library std
+    version "11"
+    language "Java"
```
Comment (dog-m) [inline src/test/testdata/lsl/generics.lsl]:
```suggestion
    version "11"
    language "unknown"
```

**Review by dog-m — CHANGES_REQUESTED** [review]

File: src/test/testdata/expected/lsl/AVLTreeArkTsFunSignatures.lsl (hunk 1 above)
Comment (alexeev509) [reply src/test/testdata/expected/lsl/AVLTreeArkTsFunSignatures.lsl]: Fixed

File: src/test/testdata/expected/lsl/generics.lsl (hunk 2 above)
Comment (alexeev509) [reply src/test/testdata/expected/lsl/generics.lsl]: Fixed

File: src/test/testdata/lsl/AVLTreeArkTsFunSignatures.lsl (hunk 5 above)
Comment (alexeev509) [reply src/test/testdata/lsl/AVLTreeArkTsFunSignatures.lsl]: Fixed

File: src/test/testdata/lsl/generics.lsl (hunk 6 above)
Comment (alexeev509) [reply src/test/testdata/lsl/generics.lsl]: Fixed;

File: src/test/testdata/expected/lsl/generics.lsl (hunk 3 above)
Comment (alexeev509) [reply src/test/testdata/expected/lsl/generics.lsl]: Fixed

File: src/test/testdata/expected/lsl/generics.lsl (hunk 4 above)
Comment (alexeev509) [reply src/test/testdata/expected/lsl/generics.lsl]: Fixed

File: src/main/kotlin/org/jetbrains/research/libsl/type/types.kt [hunk 7]
```diff
@@ -3,24 +3,25 @@ package org.jetbrains.research.libsl.type
...
 import org.jetbrains.research.libsl.nodes.references.TypeReference
-import org.jetbrains.research.libsl.type.Type.Companion.UNRESOLVED_TYPE_SYMBOL
 import org.jetbrains.research.libsl.utils.BackticksPolitics
 import org.jetbrains.research.libsl.utils.EntityPosition
 
 sealed interface Type : IPrinter {
     val name: String
     val isPointer: Boolean
     val context: LslContextBase
+    var typeBound: GenericTypeBound
```
Comment (dog-m) [inline src/main/kotlin/org/jetbrains/research/libsl/type/types.kt]: Why a type should have a bound?
As I understand it, "type + bound" mean "set/range of types".

File: src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt [hunk 8]
```diff
@@ -486,10 +492,27 @@ class ExpressionVisitor(
...
+                throw error("Constructor invoke can't contain WildCards")
+
+        }
+        
+        typeOfVariable?.genericReferences?.forEach {
+            if (!GenericTypeBound.EMPTY.equals(it.typeBound) || "?".equals(it.name))
+            // TODO: add for all exceptions in parser concrete places where it was appeared.
+                throw error("Type of variable which assigns to invokation of the constructor can't contain generics with wildcard int type")
+
+        }
```
Comment (dog-m) [inline src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt]: I am a bit confused. Why do we need to check the variable type during automaton construction?
I think these use-cases should be disallowed:
```dart
var foo: Collection<?> = new AutoCollection<?>(state=...);
var bar: Collection<?> = new AutoCollection<in String>(state=...);
```
but these -- allowed:
```dart
var fizz: Collection<?> = new AutoCollection<Int>(state=...);
var buzz: Array<out Number> = new AutoArray<Byte>(state=...);
```
*(note: attention to the automaton construction expression, not the variable declaration part)*

**Review by dog-m — CHANGES_REQUESTED** [review]

File: src/main/kotlin/org/jetbrains/research/libsl/type/types.kt (hunk 7 above)
Comment (alexeev509) [reply src/main/kotlin/org/jetbrains/research/libsl/type/types.kt]: It was deleted from here;

File: src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt (hunk 8 above)
Comment (alexeev509) [reply src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt]: Fixed;

**Review by dog-m — APPROVED** [review]

**Pull request merged**